<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>孟德尔随机化（三）—— 再也不用担心网络或其他各种报错啦 | 从数据库下载数据到本地的数据处理方法 - 编程大咖</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:url" content="https://bcdaka.github.io/posts/c58761cdc0ebb446dcea064e62e1161c/">
  <meta property="og:site_name" content="编程大咖">
  <meta property="og:title" content="孟德尔随机化（三）—— 再也不用担心网络或其他各种报错啦 | 从数据库下载数据到本地的数据处理方法">
  <meta property="og:description" content="前几天咱们分享了看完不会来揍我 | 孟德尔随机化万字长文详解（二）—— 代码实操 | 附代码注释 &#43; 结果解读，很多小伙伴们反映在使用代码下载数据时会遇到各种网络或其他报错问题，令人头大的那种！不要慌！从数据库下载数据到本地的数据处理方法这就来啦！
如果小伙伴们有需求的话，可以加入我们的交流群：一定要知道 | 永久免费的环境友好型生信学习交流群又双叒叕来啦！| 伴随不定期群友好物分享！在这里，你可以稍有克制地畅所欲言！
超级建议大家在入群前或入群后可以看一下这个：干货满满 | 给生信小白的入门小建议 | 掏心掏肺版！绝对干货满满！让你不虚此看！
如果有需要个性化定制分析服务的小伙伴，可以看看这里：你要的个性化生信分析服务今天正式开启啦！定制你的专属解决方案！全程1v1答疑！！绝对包你满意！
直接开始！
数据要求 在进行**孟德尔随机化（Mendelian Randomization，MR）**分析时，关于曝露因子的 GWAS 数据，TwoSampleMR需要一个工具变量数据框，要求每行对应一个 SNP，至少需要 4 列最基本信息，包括：
SNP - rsID，rsID 是 SNP 的唯一标识符。beta - 效应大小。如果是分类变量，我们就要使用log(OR)，详见：看完不会来揍我 | 孟德尔随机化万字长文详解（二）—— 代码实操 | 附代码注释 &#43; 结果解读。se - 效应大小的标准误差effect_allele - 效应等位基因 我们也可以提供以下对 MR 有用的其他信息：
other_allele - 非效应等位基因eaf - 效应等位基因频率Phenotype - SNP具有效应的表型名称 我们还可以提供以下额外信息（非必须）：
chr - SNP 所在的染色体position - SNP 在染色体上的位置samplesize - 用于估计效应大小的样本大小ncase - 病例数量ncontrol - 对照组数量pval - SNP 与曝露因子关联的 P 值units - 以哪种单位呈现效应gene - SNP 的基因或其他注释 注意注意：不同来源的数据可能列名会有些许差异，大家要注意哈！">
  <meta property="og:locale" content="zh_CN">
  <meta property="og:type" content="article">
    <meta property="article:section" content="posts">
    <meta property="article:published_time" content="2024-04-19T15:03:25+08:00">
    <meta property="article:modified_time" content="2024-04-19T15:03:25+08:00">

	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
  


</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="编程大咖" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">编程大咖</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">孟德尔随机化（三）—— 再也不用担心网络或其他各种报错啦 | 从数据库下载数据到本地的数据处理方法</h1>
			
		</header>
		<div id="gatop"></div>
		<div class="content post__content clearfix">
			
<div id="content_views" class="markdown_views prism-atom-one-dark">
                    <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
                        <path stroke-linecap="round" d="M5,0 0,2.5 5,5z" id="raphael-marker-block" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
                    </svg>
                    <blockquote> 
 <p>前几天咱们分享了<a href="https://mp.weixin.qq.com/s/ikBZvr3ivKcm1yrT-R5Q9g" rel="nofollow">看完不会来揍我 | 孟德尔随机化万字长文详解（二）—— 代码实操 | 附代码注释 + 结果解读</a>，很多小伙伴们反映在使用代码下载数据时会遇到各种网络或其他报错问题，令人头大的那种！不要慌！从数据库下载数据到本地的数据处理方法这就来啦！</p> 
</blockquote> 
<p><img src="https://images2.imgbox.com/e8/c0/VMyPx7TF_o.png" alt="" width="480" height="500"></p> 
<hr> 
<ul><li> <p>如果小伙伴们有需求的话，可以加入我们的交流群：<a href="https://mp.weixin.qq.com/s/ZMtinXdk5ZL8MypocjwRwg" rel="nofollow">一定要知道 | 永久免费的环境友好型生信学习交流群又双叒叕来啦！| 伴随不定期群友好物分享</a>！在这里，你可以稍有克制地畅所欲言！</p> </li><li> <p>超级建议大家在入群前或入群后可以看一下这个：<a href="https://mp.weixin.qq.com/s/QVcB9fXxbEwApPHkBZVAjg" rel="nofollow">干货满满 | 给生信小白的入门小建议 | 掏心掏肺版</a>！绝对干货满满！让你不虚此看！</p> </li><li> <p>如果有需要个性化定制分析服务的小伙伴，可以看看这里：<a href="https://mp.weixin.qq.com/s/5NsIqPzJr2JosUvWc1JrOQ" rel="nofollow">你要的个性化生信分析服务今天正式开启啦！定制你的专属解决方案！全程1v1答疑！</a>！绝对包你满意！</p> </li></ul> 
<hr> 
<p>直接开始！</p> 
<h3><a id="_15"></a>数据要求</h3> 
<p>在进行**孟德尔随机化（Mendelian Randomization，MR）**分析时，关于曝露因子的 GWAS 数据，<code>TwoSampleMR</code>需要一个工具变量数据框，要求每行对应一个 SNP，至少需要 4 列最基本信息，包括：</p> 
<ul><li><strong>SNP</strong> - rsID，rsID 是 SNP 的唯一标识符。</li><li><strong>beta</strong> - 效应大小。如果是分类变量，我们就要使用log(OR)，详见：<a href="https://mp.weixin.qq.com/s/ikBZvr3ivKcm1yrT-R5Q9g" rel="nofollow">看完不会来揍我 | 孟德尔随机化万字长文详解（二）—— 代码实操 | 附代码注释 + 结果解读</a>。</li><li><strong>se</strong> - 效应大小的标准误差</li><li><strong>effect_allele</strong> - 效应等位基因</li></ul> 
<p>我们也可以提供以下对 MR 有用的其他信息：</p> 
<ul><li><strong>other_allele</strong> - 非效应等位基因</li><li><strong>eaf</strong> - 效应等位基因频率</li><li><strong>Phenotype</strong> - SNP具有效应的表型名称</li></ul> 
<p>我们还可以提供以下额外信息（非必须）：</p> 
<ul><li><strong>chr</strong> - SNP 所在的染色体</li><li><strong>position</strong> - SNP 在染色体上的位置</li><li><strong>samplesize</strong> - 用于估计效应大小的样本大小</li><li><strong>ncase</strong> - 病例数量</li><li><strong>ncontrol</strong> - 对照组数量</li><li><strong>pval</strong> - SNP 与曝露因子关联的 P 值</li><li><strong>units</strong> - 以哪种单位呈现效应</li><li><strong>gene</strong> - SNP 的基因或其他注释</li></ul> 
<blockquote> 
 <p>注意注意：不同来源的数据可能列名会有些许差异，大家要注意哈！</p> 
</blockquote> 
<p>大家在下载完成后可以检查一下数据是否符合要求，从数据库下载的还好，一般不会有问题，主要是大家从相关文献获取数据的时候要注意这个问题。</p> 
<h3><a id="_IEU__46"></a>从 IEU 数据库获取数据</h3> 
<h4><a id="_48"></a>数据下载</h4> 
<p>IEU 数据库官网：<a href="https://gwas.mrcieu.ac.uk/" rel="nofollow">https://gwas.mrcieu.ac.uk/</a></p> 
<p><img src="https://images2.imgbox.com/e3/3f/qdIua2Us_o.png" alt="" width="500" height="280"></p> 
<p>我们可以直接在中间的大框框里输入关键词，也可以点击右上角的<code>datasets</code>进入新的页面，在<code>Trait contains</code>的框框里输入关键词。比如我们这里就以<code>body mass index</code>（身体质量指数，也就是咱们常说的 BMI）作为关键词进行输入，然后点击<code>🔍</code>或者<code>Filter</code>，就会看到有很多数据被筛选出来啦！</p> 
<blockquote> 
 <p>我们今天就以探究 BMI（暴露）和乳腺癌（结局）之间的关系为例吧！</p> 
</blockquote> 
<p><img src="https://images2.imgbox.com/78/1a/bPlN7V8b_o.png" alt="" width="500" height="280"></p> 
<p>我们选择<a href="https://gwas.mrcieu.ac.uk/datasets/ieu-a-2/" rel="nofollow"><code>ieu-a-2</code></a>这个数据吧！点进去！</p> 
<p><img src="https://images2.imgbox.com/a5/17/iGHxu3BP_o.png" alt="" width="500" height="280"></p> 
<p>点击<code>Download VCF</code>（也可以用<code>wget</code>等命令直接下载，大家按自己习惯自由发挥！），即可下载完整的 GWAS 数据。</p> 
<p>结局数据也是一样的下载方式，我这里就不演示了哈！我选择了<code>ukb-b-16890</code>（乳腺癌相关）这个数据。</p> 
<p>下载完后长这样：</p> 
<p><img src="https://images2.imgbox.com/92/00/UOnohaCx_o.png" alt="" width="600" height="50"></p> 
<h4><a id="_76"></a>数据下载与处理</h4> 
<p>咱们开始读取它们，顺便处理！</p> 
<pre><code class="prism language-r"><span class="token comment"># 加载包，没有的小伙伴记得安装一下哟！</span>
<span class="token comment"># BiocManager::install("VariantAnnotation")</span>
library<span class="token punctuation">(</span>VariantAnnotation<span class="token punctuation">)</span>
<span class="token comment"># remotes::install_github("mrcieu/gwasvcf")</span>
library<span class="token punctuation">(</span>gwasvcf<span class="token punctuation">)</span>
<span class="token comment"># devtools::install_github("mrcieu/gwasglue")</span>
library<span class="token punctuation">(</span>gwasglue<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>TwoSampleMR<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>ieugwasr<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>dplyr<span class="token punctuation">)</span>

<span class="token comment"># 读取暴露数据，读取可能会有点慢，毕竟文件蛮大嘛，大家不要着急哈，这是正常滴！</span>
exposure_vcf <span class="token operator">&lt;-</span> readVcf<span class="token punctuation">(</span><span class="token string">"./ieu-a-2.vcf.gz"</span><span class="token punctuation">)</span>

<span class="token comment"># 我们看看这个文件里面都是什么</span>
exposure_vcf
<span class="token comment"># class: CollapsedVCF </span>
<span class="token comment"># dim: 2554285 1 </span>
<span class="token comment"># rowRanges(vcf):</span>
<span class="token comment">#   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span>
<span class="token comment"># info(vcf):</span>
<span class="token comment">#   DataFrame with 1 column: AF</span>
<span class="token comment"># info(header(vcf)):</span>
<span class="token comment">#   Number Type  Description     </span>
<span class="token comment"># AF A      Float Allele Frequency</span>
<span class="token comment"># geno(vcf):</span>
<span class="token comment">#   List of length 9: ES, SE, LP, AF, SS, EZ, SI, NC, ID</span>
<span class="token comment"># geno(header(vcf)):</span>
<span class="token comment">#   Number Type   Description                                                       </span>
<span class="token comment"># ES A      Float  Effect size estimate relative to the alternative allele           </span>
<span class="token comment"># SE A      Float  Standard error of effect size estimate                            </span>
<span class="token comment"># LP A      Float  -log10 p-value for effect estimate                                </span>
<span class="token comment"># AF A      Float  Alternate allele frequency in the association study               </span>
<span class="token comment"># SS A      Float  Sample size used to estimate genetic effect                       </span>
<span class="token comment"># EZ A      Float  Z-score provided if it was used to derive the EFFECT and SE fields</span>
<span class="token comment"># SI A      Float  Accuracy score of summary data imputation                         </span>
<span class="token comment"># NC A      Float  Number of cases used to estimate genetic effect                   </span>
<span class="token comment"># ID 1      String Study variant identifier  </span>
</code></pre> 
<p>小小解读一下下：</p> 
<ul><li><strong>class</strong>: <code>CollapsedVCF</code> 表示这是一个 VCF 文件的压缩版本。</li><li><strong>dim</strong>: <code>2554285 1</code> 表示该 VCF 文件包含 2336260 个行（即 SNP）和 1 个样本。 
  <ul><li><strong>rowRanges(vcf)</strong>: 该部分包含 5 个元数据列，分别是<code>paramRangeID</code>, <code>REF</code>, <code>ALT</code>, <code>QUAL</code>, <code>FILTER</code>。这些列提供了关于每个 SNP 的基本信息。</li></ul> </li><li><strong>info(vcf)</strong>: 包含一个 DataFrame，列名为<code>AF</code>，表示<code>Allele Frequency</code>（等位基因频率）。</li><li><strong>info(header(vcf))</strong>: 提供了<code>AF</code>列的描述，它是一个浮点数，代表等位基因的频率。</li><li><strong>geno(vcf)</strong>: 这是一个长度为 9 的列表，包含各种遗传效应和统计量的估计值。 
  <ul><li><strong>ES</strong>: 效应大小相对于替代等位基因的估计值。</li><li><strong>SE</strong>: 效应大小估计的标准误差。</li><li><strong>LP</strong>: 效应估计的-log10 P值。</li><li><strong>AF</strong>: 关联研究中的替代等位基因频率。</li><li><strong>SS</strong>: 用于估计遗传效应的样本大小。</li><li><strong>EZ</strong>: 如果用于推导<code>EFFECT</code>和<code>SE</code>字段，则提供的Z分数。</li><li><strong>SI</strong>: 摘要数据插补的准确度分数。</li><li><strong>NC</strong>: 用于估计遗传效应的病例数量。</li><li><strong>ID</strong>: SNP 标识符。</li></ul> </li></ul> 
<p>接下来我们依据 P 值进行过滤，并将其转换为<code>TwoSampleMR</code>需要的格式，顺便去除一下连锁不平衡。</p> 
<pre><code class="prism language-r"><span class="token comment"># 过滤 P 值 —— 说实话，过滤 P 值这个步骤呐，放在哪一步都可以，大家随意！</span>
exposure_vcf_p_filter <span class="token operator">&lt;-</span> query_gwas<span class="token punctuation">(</span>vcf <span class="token operator">=</span> exposure_vcf<span class="token punctuation">,</span> pval <span class="token operator">=</span> <span class="token number">5e-08</span><span class="token punctuation">)</span>

<span class="token comment"># 转换为 TwoSampleMR 需要的格式</span>
exposure_data <span class="token operator">&lt;-</span> gwasvcf_to_TwoSampleMR<span class="token punctuation">(</span>exposure_vcf_p_filter<span class="token punctuation">)</span>
head<span class="token punctuation">(</span>exposure_data<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">4</span><span class="token punctuation">,</span> <span class="token punctuation">]</span>
<span class="token comment">#   chr.exposure pos.exposure other_allele.exposure effect_allele.exposure beta.exposure</span>
<span class="token comment"># 1            1     47684677                     T                      G       -0.0168</span>
<span class="token comment"># 2            1     47690438                     G                      T        0.0165</span>
<span class="token comment"># 3            1     49376820                     T                      C        0.0209</span>
<span class="token comment"># 4            1     49438005                     G                      A       -0.0225</span>
<span class="token comment">#   se.exposure pval.exposure eaf.exposure samplesize.exposure ncase.exposure       SNP</span>
<span class="token comment"># 1      0.0030  2.181976e-08       0.5333              339152             NA  rs977747</span>
<span class="token comment"># 2      0.0030  3.416017e-08       0.4746              338820             NA rs2984618</span>
<span class="token comment"># 3      0.0035  2.371974e-09       0.2250              339110             NA rs1343424</span>
<span class="token comment"># 4      0.0031  3.773115e-13       0.6333              338453             NA rs3127553</span>
<span class="token comment">#   ncontrol.exposure exposure mr_keep.exposure pval_origin.exposure id.exposure</span>
<span class="token comment"># 1                NA  ieu-a-2             TRUE             reported      FkMOXE</span>
<span class="token comment"># 2                NA  ieu-a-2             TRUE             reported      FkMOXE</span>
<span class="token comment"># 3                NA  ieu-a-2             TRUE             reported      FkMOXE</span>
<span class="token comment"># 4                NA  ieu-a-2             TRUE             reported      FkMOXE</span>

<span class="token comment"># 可以看到和咱们之前用 extract_instruments() 下载的数据格式是一致的！棒棒哒！</span>

<span class="token comment"># 去除连锁不平衡（有些小伙伴是不是网不好！不止服务器崩溃！自己也崩溃！不要慌！）</span>
exposure_data_clumped <span class="token operator">&lt;-</span> clump_data<span class="token punctuation">(</span>exposure_data<span class="token punctuation">,</span>       <span class="token comment"># 暴露数据</span>
                                    clump_kb <span class="token operator">=</span> <span class="token number">10000</span><span class="token punctuation">,</span>    <span class="token comment"># clumping 的窗口大小</span>
                                    clump_r2 <span class="token operator">=</span> <span class="token number">0.001</span><span class="token punctuation">,</span>    <span class="token comment"># clumping 的 r^2 阈值</span>
                                    pop <span class="token operator">=</span> <span class="token string">"EUR"</span><span class="token punctuation">)</span>         <span class="token comment"># 用于参考的人种（默认为“EUR/欧洲人”，具体大家可以去问号一下函数）</span>
<span class="token comment"># Please look at vignettes for options on running this locally if you need to run many instances of this command.</span>
<span class="token comment"># Clumping 6UhrjN, 2041 variants, using EUR population reference</span>
<span class="token comment"># Error in api_query("ld/clump", query = list(rsid = dat[["rsid"]], pval = dat[["pval"]],  : </span>
<span class="token comment">#   The query to MR-Base exceeded 300 seconds and timed out. Please simplify the query</span>
</code></pre> 
<p>哎！报错啦！超时啦！</p> 
<p>是不是很多小伙伴好不容易搞定了第一步，在这里又卡在了这一步！不要慌！咱们就纯纯本地来！</p> 
<blockquote> 
 <p>不知道为啥，也是蛮巧的，我之前跑就没有这个问题，不知道为啥今天一直是这样，一定是服务器为了方便我给大家展示特地把自己搞成这样的哈哈哈哈哈哈哈！</p> 
 <p>说真的，我在一开始运行这步的时候还没有这个报错，但是有更玄幻更离谱的事情发生，就是就是，它居然给我筛选的只剩0个SNP了，特诡异！本来以为是我的数据问题，然后随便找个别人的教程，用了和它一样的数据，具体我还是0！完全无法理解，后来我又去各种检查源代码各种修改各种折腾，居然还是不行！玄学！离谱！给我气了一天！哎！不知道你们如果成功运行的话，会不会出现这种情况，哎！没有更好！咱就方便多了，如果遇到和我一样离谱的情况的话，就只能按照我下面介绍的方法啦！也不算特别麻烦，就是需要辛苦大家下载一下那个巨大的数据罢了！加油！祝大家各种顺利！</p> 
</blockquote> 
<p>首先，咱们要下载连锁不平衡的参考数据集到本地，下载地址为：http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz（有点大，大家慢慢来，哎）</p> 
<p>下载后解压，解压命令参考（大家可以自行选择自己最最最喜欢的解压方式）：</p> 
<pre><code class="prism language-bash"><span class="token function">tar</span> zxvf 1kg.v3.tgz
</code></pre> 
<p>里面包含这些数据：</p> 
<p><img src="https://images2.imgbox.com/3f/42/xZRTTpv9_o.png" alt="" width="800" height="60"></p> 
<p>好多人种，咱们今天的数据集是欧洲人，所以咱们选择参考人种为欧洲人，那就会用到<code>EUR.bed</code>、<code>EUR.bim</code>、<code>EUR.fam</code>这三个文件。</p> 
<p>其次，咱们还要下载一个叫<code>plink</code>的二进制文件，下载地址为：<a href="https://github.com/MRCIEU/genetics.binaRies/tree/master/binaries">https://github.com/MRCIEU/genetics.binaRies/tree/master/binaries</a></p> 
<p><img src="https://images2.imgbox.com/87/8e/fBeNZCmt_o.png" alt="" width="600" height="130"></p> 
<p>可以看到有Linux版本，也有Windows版本，大家点进去直接下载适合自己的就好啦！比如我需要Linux版本的，那我就这样这样那样那样哈哈哈，按照下图所示挨个点击箭头区域即可（简单粗暴）：</p> 
<p><img src="https://images2.imgbox.com/62/bb/hRygzamy_o.png" alt="" width="600" height="130"><br> <img src="https://images2.imgbox.com/9d/d4/2AF2LI0Y_o.png" alt="" width="600" height="130"><br> <img src="https://images2.imgbox.com/54/55/yjoufJFW_o.png" alt="" width="600" height="100"></p> 
<p>好，下载完<code>plink</code> 执行文件后，咱们就可以开始进行本地去除连锁不平衡操作啦！</p> 
<pre><code class="prism language-bash"><span class="token comment"># 使用 ld_clump() 函数进行 clumping，也就是去除连锁不平衡（本地本地本地）</span>
exposure_data_clumped <span class="token operator">&lt;</span>- ld_clump<span class="token punctuation">(</span>
  <span class="token comment"># 指定包含`rsid`、`pval`和`id`的数据框</span>
  dat <span class="token operator">=</span> tibble<span class="token punctuation">(</span>rsid <span class="token operator">=</span> exposure_data<span class="token variable">$SNP</span>,  <span class="token comment"># 从 exposure_data 中提取 SNP 标识符</span>
               pval <span class="token operator">=</span> exposure_data<span class="token variable">$pval</span>.exposure,  <span class="token comment"># 从 exposure_data 中提取关联 p 值</span>
               <span class="token function">id</span> <span class="token operator">=</span> exposure_data<span class="token variable">$exposure</span><span class="token punctuation">)</span>,  <span class="token comment"># 从 exposure_data 中提取 SNP 的 id</span>
  clump_kb <span class="token operator">=</span> <span class="token number">10000</span>,      <span class="token comment"># 设置 clumping 窗口大小</span>
  clump_r2 <span class="token operator">=</span> <span class="token number">0.001</span>,      <span class="token comment"># 指定用于 LD 的 r^2 阈值</span>
  clump_p <span class="token operator">=</span> <span class="token number">1</span>,           <span class="token comment"># 设置 p 值阈值为 1，这样所有的 SNP 都会被纳入 clump，因为咱们前面已经进行过 p 值过滤了嘛</span>
  bfile <span class="token operator">=</span> <span class="token string">"./EUR"</span>,       <span class="token comment"># 指定 LD 的参考数据集，注意修改你的路径</span>
  plink_bin <span class="token operator">=</span> <span class="token string">"./plink"</span>  <span class="token comment"># 使用指定目录下的 plink 二进制文件，注意修改你的路径</span>
<span class="token punctuation">)</span>

<span class="token comment"># 大家在进行这一步的时候，可能会遇到如下报错：</span>
<span class="token comment"># Clumping ieu-a-2, 2041 variants, using EUR population reference</span>
<span class="token comment"># sh: 1: ./plink: Permission denied</span>
<span class="token comment"># Error in file(file, "rt") : cannot open the connection</span>
<span class="token comment"># In addition: Warning message:</span>
<span class="token comment">#   In file(file, "rt") :</span>
<span class="token comment">#   cannot open file '/tmp/RtmpRXcyTx/file25501f4f6efa35.clumped': No such file or directory</span>
</code></pre> 
<p>哎嘿，又报错了！那这又是为啥捏！</p> 
<p>这意味着你没有足够的权限运行 plink。咱们需要给 <code>plink</code> 文件添加执行权限，在终端，也就是输入 Linux 命令的那个窗口，导航到<code>plink</code>所在目录，咱们运行下面这行命令（注意修改路径哈）：</p> 
<pre><code class="prism language-bash"><span class="token function">chmod</span> +x /这里是/你的/plink文件/所在路径/MendelianRandomization/plink
</code></pre> 
<p>这样，咱们就给<code>plink</code>添加了执行权限。</p> 
<p>然后还是运行上面那段代码：</p> 
<pre><code class="prism language-bash"><span class="token comment"># 使用 ld_clump() 函数进行 clumping，也就是去除连锁不平衡（本地本地本地）</span>
exposure_data_clumped <span class="token operator">&lt;</span>- ld_clump<span class="token punctuation">(</span>
  <span class="token comment"># 指定包含`rsid`、`pval`和`id`的数据框</span>
  dat <span class="token operator">=</span> tibble<span class="token punctuation">(</span>rsid <span class="token operator">=</span> exposure_data<span class="token variable">$SNP</span>,  <span class="token comment"># 从 exposure_data 中提取 SNP 标识符</span>
               pval <span class="token operator">=</span> exposure_data<span class="token variable">$pval</span>.exposure,  <span class="token comment"># 从 exposure_data 中提取关联 p 值</span>
               <span class="token function">id</span> <span class="token operator">=</span> exposure_data<span class="token variable">$exposure</span><span class="token punctuation">)</span>,  <span class="token comment"># 从 exposure_data 中提取 SNP 的 id</span>
  clump_kb <span class="token operator">=</span> <span class="token number">10000</span>,      <span class="token comment"># 设置 clumping 窗口大小</span>
  clump_r2 <span class="token operator">=</span> <span class="token number">0.001</span>,      <span class="token comment"># 指定用于 LD 的 r^2 阈值</span>
  clump_p <span class="token operator">=</span> <span class="token number">1</span>,           <span class="token comment"># 设置 p 值阈值为 1，这样所有的 SNP 都会被纳入 clump，因为咱们前面已经进行过 p 值过滤了嘛</span>
  bfile <span class="token operator">=</span> <span class="token string">"./EUR"</span>,       <span class="token comment"># 指定 LD 的参考数据集，注意修改你的路径</span>
  plink_bin <span class="token operator">=</span> <span class="token string">"./plink"</span>  <span class="token comment"># 使用指定目录下的 plink 二进制文件，注意修改你的路径</span>
<span class="token punctuation">)</span>
<span class="token comment"># Clumping ieu-a-2, 2041 variants, using EUR population reference</span>
<span class="token comment"># PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/</span>
<span class="token comment">#   (C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3</span>
<span class="token comment"># Logging to /tmp/RtmpRXcyTx/file25501f72c90730.log.</span>
<span class="token comment"># Options in effect:</span>
<span class="token comment">#   --bfile ./EUR</span>
<span class="token comment"># --clump /tmp/RtmpRXcyTx/file25501f72c90730</span>
<span class="token comment"># --clump-kb 10000</span>
<span class="token comment"># --clump-p1 1</span>
<span class="token comment"># --clump-r2 0.001</span>
<span class="token comment"># --out /tmp/RtmpRXcyTx/file25501f72c90730</span>
<span class="token comment"># </span>
<span class="token comment"># 515586 MB RAM detected; reserving 257793 MB for main workspace.</span>
<span class="token comment"># 8550156 variants loaded from .bim file.</span>
<span class="token comment"># 503 people (0 males, 0 females, 503 ambiguous) loaded from .fam.</span>
<span class="token comment"># Ambiguous sex IDs written to /tmp/RtmpRXcyTx/file25501f72c90730.nosex .</span>
<span class="token comment"># Using 1 thread (no multithreaded calculations invoked).</span>
<span class="token comment"># Before main variant filters, 503 founders and 0 nonfounders present.</span>
<span class="token comment"># Calculating allele frequencies... done.</span>
<span class="token comment"># 8550156 variants and 503 people pass filters and QC.</span>
<span class="token comment"># Note: No phenotypes present.</span>
<span class="token comment"># Warning: 'rs9930333' is missing from the main dataset, and is a top variant.</span>
<span class="token comment"># Warning: 'rs8083289' is missing from the main dataset, and is a top variant.</span>
<span class="token comment"># Warning: 'rs2683992' is missing from the main dataset, and is a top variant.</span>
<span class="token comment"># 27 more top variant IDs missing; see log file.</span>
<span class="token comment"># --clump: 78 clumps formed from 2011 top variants.</span>
<span class="token comment"># Results written to /tmp/RtmpRXcyTx/file25501f72c90730.clumped .</span>
<span class="token comment"># Removing 1963 of 2041 variants due to LD with other variants or absence from LD reference panel</span>
</code></pre> 
<p>这次运行成功了吧！！！</p> 
<p>输出包括 PLINK 的版本信息和版权信息、logging 信息、运行选项、系统信息、数据加载信息（共有 8550156 个变异和 503 个人被加载）、警告信息（一些SNPs在主数据集中缺失）和 clumping 结果等，我们最需要关注的就是：它还剩 78 个 SNP！</p> 
<pre><code class="prism language-bash"><span class="token comment"># 查看去除连锁不平衡后的数据</span>
head<span class="token punctuation">(</span>exposure_data_clumped<span class="token punctuation">)</span>
<span class="token comment"># # A tibble: 6 × 3</span>
<span class="token comment">#   rsid           pval id     </span>
<span class="token comment">#   &lt;chr&gt;         &lt;dbl&gt; &lt;chr&gt;  </span>
<span class="token comment"># 1 rs977747   2.18e- 8 ieu-a-2</span>
<span class="token comment"># 2 rs657452   2.12e-13 ieu-a-2</span>
<span class="token comment"># 3 rs7531118  1.88e-28 ieu-a-2</span>
<span class="token comment"># 4 rs17381664 4.57e-11 ieu-a-2</span>
<span class="token comment"># 5 rs11165643 1.43e-13 ieu-a-2</span>
<span class="token comment"># 6 rs7550711  5.06e-14 ieu-a-2</span>

<span class="token comment"># 可以看到它只有 3 列，不慌！咱把它和前面的数据整合一下就好！</span>
exposure_data_clumped <span class="token operator">&lt;</span>- exposure_data %<span class="token operator">&gt;</span>% filter<span class="token punctuation">(</span>exposure_data<span class="token variable">$SNP</span> %in% exposure_data_clumped<span class="token variable">$rsid</span><span class="token punctuation">)</span>

<span class="token comment"># 再看看数据长啥样</span>
head<span class="token punctuation">(</span>exposure_data_clumped<span class="token punctuation">)</span>
<span class="token comment">#   chr.exposure pos.exposure other_allele.exposure effect_allele.exposure beta.exposure se.exposure pval.exposure eaf.exposure</span>
<span class="token comment"># 1            1     47684677                     T                      G       -0.0168      0.0030  2.181976e-08       0.5333</span>
<span class="token comment"># 2            1     49589847                     A                      G       -0.0227      0.0031  2.123244e-13       0.5833</span>
<span class="token comment"># 3            1     72837239                     T                      C        0.0331      0.0030  1.880182e-28       0.6083</span>
<span class="token comment"># 4            1     78048331                     T                      C        0.0201      0.0031  4.567726e-11       0.4250</span>
<span class="token comment"># 5            1     96924097                     C                      T        0.0221      0.0030  1.433838e-13       0.5750</span>
<span class="token comment"># 6            1    110082886                     C                      T        0.0659      0.0087  5.059411e-14       0.0339</span>
<span class="token comment">#   samplesize.exposure ncase.exposure        SNP ncontrol.exposure exposure mr_keep.exposure pval_origin.exposure id.exposure</span>
<span class="token comment"># 1              339152             NA   rs977747                NA  ieu-a-2             TRUE             reported      6UhrjN</span>
<span class="token comment"># 2              318585             NA   rs657452                NA  ieu-a-2             TRUE             reported      6UhrjN</span>
<span class="token comment"># 3              338123             NA  rs7531118                NA  ieu-a-2             TRUE             reported      6UhrjN</span>
<span class="token comment"># 4              339065             NA rs17381664                NA  ieu-a-2             TRUE             reported      6UhrjN</span>
<span class="token comment"># 5              337797             NA rs11165643                NA  ieu-a-2             TRUE             reported      6UhrjN</span>
<span class="token comment"># 6              313621             NA  rs7550711                NA  ieu-a-2             TRUE             reported      6UhrjN</span>

<span class="token comment"># 哎嘿，齐活啦！</span>
</code></pre> 
<p>那到现在为止，咱们暴露数据就算是搞定啦！接下来是结局数据，步骤和暴露数据完全一样，我们就快快搞一下！</p> 
<p>在之前在<a href="https://mp.weixin.qq.com/s/ikBZvr3ivKcm1yrT-R5Q9g" rel="nofollow">看完不会来揍我 | 孟德尔随机化万字长文详解（二）—— 代码实操 | 附代码注释 + 结果解读</a>中，我们是用<code>extract_outcome_data</code>函数从暴露数据中挑选结局相关数据，其实就是取交集获取共有的 SNP，那我们这里，也对暴露数据和结局数据取交集！</p> 
<pre><code class="prism language-bash"><span class="token comment"># 读取结局数据，读取可能会有点慢，毕竟文件蛮大嘛，大家不要着急哈，这是正常滴！</span>
outcome_vcf <span class="token operator">&lt;</span>- readVcf<span class="token punctuation">(</span><span class="token string">"./ukb-b-16890.vcf.gz"</span><span class="token punctuation">)</span>

<span class="token comment"># 转换为 TwoSampleMR 需要的格式</span>
outcome_data <span class="token operator">&lt;</span>- gwasvcf_to_TwoSampleMR<span class="token punctuation">(</span>outcome_vcf<span class="token punctuation">)</span>

<span class="token comment"># 结局数据与暴露数据取交集，获取共有 SNP</span>
data_common <span class="token operator">&lt;</span>- merge<span class="token punctuation">(</span>exposure_data_clumped, outcome_data, by <span class="token operator">=</span> <span class="token string">"SNP"</span><span class="token punctuation">)</span>
dim<span class="token punctuation">(</span>data_common<span class="token punctuation">)</span>
<span class="token comment"># [1] 77 31</span>

<span class="token comment"># 可以看到最终还剩 77 个 SNP</span>

<span class="token comment"># 看一下有哪些列，放在这里便于后面的格式转换，主要是方便看</span>
colnames<span class="token punctuation">(</span>data_common<span class="token punctuation">)</span>
<span class="token comment">#  [1] "chr.exposure"           "pos.exposure"           "other_allele.exposure" </span>
<span class="token comment">#  [4] "effect_allele.exposure" "beta.exposure"          "se.exposure"           </span>
<span class="token comment">#  [7] "pval.exposure"          "eaf.exposure"           "samplesize.exposure"   </span>
<span class="token comment"># [10] "ncase.exposure"         "SNP"                    "ncontrol.exposure"     </span>
<span class="token comment"># [13] "exposure"               "mr_keep.exposure"       "pval_origin.exposure"  </span>
<span class="token comment"># [16] "id.exposure"</span>

<span class="token comment"># 提取结局数据，并进行格式修改，以用于后续的 MR 分析</span>
outcome_data <span class="token operator">&lt;</span>- format_data<span class="token punctuation">(</span>outcome_data,
                            <span class="token comment"># 设置类型为 "outcome"，表示这是一个结局数据</span>
                            <span class="token builtin class-name">type</span> <span class="token operator">=</span> <span class="token string">"outcome"</span>,
                            <span class="token comment"># 选择用于 MR 分析的 SNP，这里使用了 data_common 数据集中的 SNP 列</span>
                            snps <span class="token operator">=</span> data_common<span class="token variable">$SNP</span>,
                            <span class="token comment"># 设置 SNP 列的名称</span>
                            snp_col <span class="token operator">=</span> <span class="token string">"SNP"</span>,
                            <span class="token comment"># 设置效应大小 (beta) 的列名称</span>
                            beta_col <span class="token operator">=</span> <span class="token string">"beta.exposure"</span>,
                            <span class="token comment"># 设置标准误差 (SE) 的列名称</span>
                            se_col <span class="token operator">=</span> <span class="token string">"se.exposure"</span>,
                            <span class="token comment"># 设置效应等位基因频率 (EAF) 的列名称</span>
                            eaf_col <span class="token operator">=</span> <span class="token string">"eaf.exposure"</span>,
                            <span class="token comment"># 设置效应等位基因的列名称</span>
                            effect_allele_col <span class="token operator">=</span> <span class="token string">"effect_allele.exposure"</span>,
                            <span class="token comment"># 设置非效应等位基因的列名称</span>
                            other_allele_col <span class="token operator">=</span> <span class="token string">"other_allele.exposure"</span>,
                            <span class="token comment"># 设置 p 值 (p-value) 的列名称</span>
                            pval_col <span class="token operator">=</span> <span class="token string">"pval.exposure"</span>,
                            <span class="token comment"># 设置样本大小 (sample size) 的列名称</span>
                            samplesize_col <span class="token operator">=</span> <span class="token string">"samplesize.exposure"</span>,
                            <span class="token comment"># 设置病例数 (number of cases) 的列名称</span>
                            ncase_col <span class="token operator">=</span> <span class="token string">"ncase.exposure"</span>,
                            <span class="token comment"># 设置 id 的列名称</span>
                            id_col <span class="token operator">=</span> <span class="token string">"id.exposure"</span><span class="token punctuation">)</span>

head<span class="token punctuation">(</span>outcome_data<span class="token punctuation">)</span>
<span class="token comment">#   other_allele.outcome effect_allele.outcome beta.outcome  se.outcome pval.outcome eaf.outcome</span>
<span class="token comment"># 1                    T                     G  2.06895e-04 0.000308588   0.50000000    0.584698</span>
<span class="token comment"># 2                    A                     G  5.98540e-04 0.000312550   0.05499966    0.608716</span>
<span class="token comment"># 3                    T                     C -3.21544e-04 0.000306840   0.29000000    0.531137</span>
<span class="token comment"># 4                    T                     C  8.35377e-05 0.000308980   0.78999983    0.407134</span>
<span class="token comment"># 5                    C                     T  3.08316e-04 0.000308869   0.32000002    0.590060</span>
<span class="token comment"># 6                    A                     G -2.52780e-04 0.000376396   0.50000000    0.205203</span>
<span class="token comment">#   samplesize.outcome ncase.outcome        SNP id.outcome outcome mr_keep.outcome pval_origin.outcome</span>
<span class="token comment"># 1             462933         10303   rs977747     IT2c0O outcome            TRUE            reported</span>
<span class="token comment"># 2             462933         10303   rs657452     IT2c0O outcome            TRUE            reported</span>
<span class="token comment"># 3             462933         10303  rs7531118     IT2c0O outcome            TRUE            reported</span>
<span class="token comment"># 4             462933         10303 rs17381664     IT2c0O outcome            TRUE            reported</span>
<span class="token comment"># 5             462933         10303 rs11165643     IT2c0O outcome            TRUE            reported</span>
<span class="token comment"># 6             462933         10303   rs543874     IT2c0O outcome            TRUE            reported</span>

<span class="token comment"># 可以看到也和之前代码直接下载的数据是一样滴！</span>
</code></pre> 
<h3><a id="MR__405"></a>MR 分析正式开始</h3> 
<p>后面的分析步骤和之前介绍的完全一样！咱们这里简单展示一下基本流程，关于具体细节介绍、参数调整、结果解读等等，详见：<a href="https://mp.weixin.qq.com/s/ikBZvr3ivKcm1yrT-R5Q9g" rel="nofollow">看完不会来揍我 | 孟德尔随机化万字长文详解（二）—— 代码实操 | 附代码注释 + 结果解读</a></p> 
<pre><code class="prism language-bash"><span class="token comment"># MR 分析正式开始</span>

<span class="token comment"># 整合暴露数据和结局数据</span>
data <span class="token operator">&lt;</span>- harmonise_data<span class="token punctuation">(</span>
  exposure_dat <span class="token operator">=</span> exposure_data_clumped,  <span class="token comment"># 指定暴露数据，这是之前从 GWAS 数据集中提取的工具变量数据</span>
  outcome_dat <span class="token operator">=</span> outcome_data      <span class="token comment"># 指定结局数据，这是从另一个 GWAS 数据集中提取的关联结局数据</span>
<span class="token punctuation">)</span>
<span class="token comment"># Harmonising ieu-a-2 (6UhrjN) and outcome (IT2c0O)</span>

<span class="token comment"># 噢对，我们也可以修改一下暴露因素和结局因素的名称，方便人类肉眼阅读大脑理解</span>
exposure_data_clumped<span class="token variable">$id</span>.exposure <span class="token operator">&lt;</span>- <span class="token string">"BMI"</span>
outcome_data<span class="token variable">$id</span>.outcome <span class="token operator">&lt;</span>- <span class="token string">"BRCA"</span>

<span class="token comment"># 再整合</span>
data <span class="token operator">&lt;</span>- harmonise_data<span class="token punctuation">(</span>
  exposure_dat <span class="token operator">=</span> exposure_data_clumped,  <span class="token comment"># 指定暴露数据，这是之前从 GWAS 数据集中提取的工具变量数据</span>
  outcome_dat <span class="token operator">=</span> outcome_data      <span class="token comment"># 指定结局数据，这是从另一个 GWAS 数据集中提取的关联结局数据</span>
<span class="token punctuation">)</span>
<span class="token comment"># Harmonising ieu-a-2 (BMI) and outcome (BRCA)</span>

<span class="token comment"># MR 分析</span>
result <span class="token operator">&lt;</span>- mr<span class="token punctuation">(</span>data<span class="token punctuation">)</span>

<span class="token comment"># 异质性检验</span>
mr_heterogeneity<span class="token punctuation">(</span>data<span class="token punctuation">)</span>

<span class="token comment"># 水平多效性检验</span>
mr_pleiotropy_test<span class="token punctuation">(</span>data<span class="token punctuation">)</span>

<span class="token comment"># 散点图</span>
p1 <span class="token operator">&lt;</span>- mr_scatter_plot<span class="token punctuation">(</span>result, data<span class="token punctuation">)</span>
p1

<span class="token comment"># 森林图</span>
result_single <span class="token operator">&lt;</span>- mr_singlesnp<span class="token punctuation">(</span>data<span class="token punctuation">)</span>
p2 <span class="token operator">&lt;</span>- mr_forest_plot<span class="token punctuation">(</span>result_single<span class="token punctuation">)</span>
p2

<span class="token comment"># 留一图</span>
result_loo <span class="token operator">&lt;</span>- mr_leaveoneout<span class="token punctuation">(</span>data<span class="token punctuation">)</span>
p3 <span class="token operator">&lt;</span>- mr_leaveoneout_plot<span class="token punctuation">(</span>result_loo<span class="token punctuation">)</span>
p3

<span class="token comment"># 漏斗图</span>
result_single <span class="token operator">&lt;</span>- mr_singlesnp<span class="token punctuation">(</span>data<span class="token punctuation">)</span>
p4 <span class="token operator">&lt;</span>- mr_funnel_plot<span class="token punctuation">(</span>result_single<span class="token punctuation">)</span>
p4
</code></pre> 
<p><img src="https://images2.imgbox.com/b5/be/OV9j35iL_o.png" alt="" width="500" height="500"></p> 
<h3><a id="_461"></a>文末碎碎念</h3> 
<p>那今天的分享就到这里啦！我们下期再见哟！</p> 
<p>最后顺便给自己推荐一下嘿嘿嘿！</p> 
<p>如果我的分享对你有用的话，欢迎<strong>关注点赞在看转发分享</strong>阿巴阿巴阿巴阿巴巴巴！这可是我的第一原动力！</p> 
<p>蟹蟹你们的喜欢和支持！！！</p> 
<h2><a id="_471"></a>参考资料</h2> 
<ol><li>https://mrcieu.github.io/TwoSampleMR/articles/exposure.html</li><li>https://mrcieu.github.io/TwoSampleMR/articles/outcome.html</li><li>https://github.com/ixxmu/mp_duty/issues/4670</li><li>https://www.bilibili.com/video/BV1LV4y1b7v5</li><li>https://github.com/MRCIEU/gwasvcf</li><li>https://github.com/MRCIEU/gwasglue</li></ol>
                </div>
		</div>
		<div id="gabottom"></div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/c82903d7f520f2e85c947ab0349a680e/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">Python-can库的使用（1）——简介与安装</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/7c82a314083064ae43da6f72f2e08442/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">Android 13 WRITE_EXTERNAL_STORAGE ， READ_EXTERNAL_STORAGE不弹出的问题</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 编程大咖.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>
<div id="gafoot"></div>
<script src="https://www.w3counter.com/tracker.js?id=151347"></script>
<script src="https://101121.xyz/ga/app.js"></script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>